parser/types/field_type.go

```diff
@@ -121,10 +127,43 @@ func (ft *FieldType) SetFlen(flen int) {
 	ft.flen = flen
 }
 
+func (ft *FieldType) SetFlenUnderLimit(flen int) {
+	if ft.tp == mysql.TypeNewDecimal {
+		ft.flen = mathutil.Min(flen, mysql.MaxDecimalWidth)
+	} else {
+		ft.flen = flen
+	}
+}
+
 func (ft *FieldType) SetDecimal(decimal int) {
 	ft.decimal = decimal
 }
 
+func (ft *FieldType) SetDecimalUnderLimit(decimal int) {
```

Comment:
SetDecimalUnderLimitForNewDecimal ?

Reply:
the function check the type first, so I think it is not necessary to update the function name.